File: src/main/python/composed_user_knot.py
from services import UserService as USER_SERVICE
from country_knot import CountryKnot


class ComposedUserKnot(object):

    def __init__(self, user_id):
        self.user_id = user_id
        self._cached_user = None
        self._cached_country_knot = None

    @property
    def user(self):
        if self._cached_user is None:
            self._cached_user = USER_SERVICE.get_user_by_id(self.user_id)

    @property
    def _country_knot(self):
        if self._cached_country_knot is None:
            self._cached_country_knot = CountryKnot(self.user.coutry_id)
        return self._cached_country_knot

    @property
    def country(self):
        return self._country_knot.country

    @property
    def currency(self):
        return self._country_knot.currency

File: src/main/java/org/kambanaria/knots/examples/ComposedUserKnot.java
package org.kambanaria.knots.examples;

import org.kambanaria.knots.microservices_models.Country;
import org.kambanaria.knots.microservices_models.Currency;
import org.kambanaria.knots.microservices_models.User;
import static org.kambanaria.knots.services.StaticRegistry.*;

public class ComposedUserKnot {

    private final int userId;
    private User user;
    private CountryKnot countryKnot;

    public ComposedUserKnot(int userId) {
        this.userId = userId;
    }

    public User getUser() {
        if (user == null) {
            user = USER_SERVICE.getUserById(userId);
        }
        return user;
    }

    private CountryKnot getCountryKnot() {
        if (countryKnot == null) {
            countryKnot = new CountryKnot(getUser().getCountryId());
        }
        return countryKnot;
    }

    public Country getCountry() {
        return getCountryKnot().getCountry();
    }

    public Currency getCurrency() {
        return getCountryKnot().getCurrency();
    }
}
